opm/material/fluidsystems/blackoilpvt/GasPvtThermal.hpp

```diff
@@ -222,16 +241,61 @@ class GasPvtThermal
     template <class Evaluation>
     Evaluation internalEnergy(unsigned regionIdx,
                               const Evaluation& temperature,
-                              const Evaluation&,
-                              const Evaluation&) const
+                              const Evaluation& pressure,
+                              const Evaluation& Rv) const
     {
         if (!enableInternalEnergy_)
-            throw std::runtime_error("Requested the internal energy of oil but it is disabled");
+             throw std::runtime_error("Requested the internal energy of gas but it is disabled");
 
-        // compute the specific internal energy for the specified tempature. We use linear
-        // interpolation here despite the fact that the underlying heat capacities are
-        // piecewise linear (which leads to a quadratic function)
-        return internalEnergyCurves_[regionIdx].eval(temperature, /*extrapolate=*/true);
+        bool enableJouleThomson_  = gasJT_[regionIdx];
+
+        if (!enableJouleThomson_) {
+            // compute the specific internal energy for the specified tempature. We use linear
+            // interpolation here despite the fact that the underlying heat capacities are
+            // piecewise linear (which leads to a quadratic function)
+            return internalEnergyCurves_[regionIdx].eval(temperature, /*extrapolate=*/true);
+        }
+        else {
+            Evaluation Tref = gasdentRefTemp_[regionIdx];
+            Evaluation Pref = gasJTRefPres_[regionIdx]; 
+            Scalar JTC = gasJTC_[regionIdx]; // JTC = 0: JTC calculated
+
+            Evaluation invB = inverseFormationVolumeFactor(regionIdx, temperature, pressure, Rv);
+            const Scalar hVap = 480.6e3; // [J / kg]
+            Evaluation Cp = (internalEnergyCurves_[regionIdx].eval(temperature, /*extrapolate=*/true) - hVap)/temperature;
+            Evaluation density = invB * gasReferenceDensity(regionIdx);
+
+            Evaluation enthalpyPres;
+            if  (JTC != 0) {
+                enthalpyPres = -Cp * JTC * (pressure -Pref);
+            }
+            else if(enableThermalDensity_) {
+                Scalar c1T = gasdentCT1_[regionIdx];
+                Scalar c2T = gasdentCT2_[regionIdx];
+
+                Evaluation alpha = (c1T + 2 * c2T * (temperature - Tref)) /
+                    (1 + c1T  *(temperature - Tref) + c2T * (temperature - Tref) * (temperature - Tref));
+
+                const int N = 100; // value is experimental
+                Evaluation deltaP = (pressure - Pref)/N;
```

Comment:
Does it make sense to use the pressure values from the table?

Reply:
The Joule-Thomson table input, see [Joule-Thomson keywords for gas, oil and water phase #3029](https://github.com/OPM/opm-common/pull/3029), has only a single reference pressure. A user can now only specify a single Joule-Thomson Coefficient (`JTC`) if she wants to overrule the internally calculated pressure and temperature dependent `JTC`. It would make sense to me, I guess you are hinting at that, to have the ability to specify as user a 2D table for `JTC` as function of temperature and pressure but that is not (yet) implemented.